cdap-watchdog/src/main/java/co/cask/cdap/logging/gateway/handlers/LogDataOffsetCallback.java

```diff
@@ -44,8 +46,20 @@ public Object encodeSend(LogEvent logEvent) {
     String className = "";
     String simpleClassName = "";
     int lineNumber = 0;
+    Level level = event.getLevel();
+    StackTraceElement first = null;
     if (stackTraceElements.length > 0) {
-      StackTraceElement first = stackTraceElements[0];
+      first = stackTraceElements[0];
+      // If the line number from caller data is -1 and the event level is ERROR,
+      // use the line number from throwable proxy
+      if (first.getLineNumber() == -1 && level.equals(Level.ERROR)) {
+        IThrowableProxy throwableProxy = event.getThrowableProxy();
+        if (throwableProxy != null && throwableProxy.getStackTraceElementProxyArray().length > 0) {
+          first = throwableProxy.getStackTraceElementProxyArray()[0].getStackTraceElement();
+        }
+      }
+    }
+    if (first != null) {
       className = first.getClassName();
```

Comment:
Why don't you just move this `if` block inside the `if (stackTraceElements.length> 0)` block? Then you don't have the check `if (first != null)` again.